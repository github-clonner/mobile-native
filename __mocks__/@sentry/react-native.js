export const captureMessage = jest.fn();
export default {};
