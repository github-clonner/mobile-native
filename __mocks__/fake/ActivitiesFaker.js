
function load(count, container) {

  let activities = [...Array(count)].map((_, i) => {
    const code = 'activityguid' + i;
    return {
      attachment_guid: false,
      blurb: false,
      container_guid:code,
      custom_data:false,
      custom_type:false,
      rowKey: 'something' + i,
      description:"Congratulations! ",
      edited:"",
      guid:code,
      mature:false,
      time_created: "1522036284",
      ownerObj:{
        guid: "824853017709780997",
        type: "user",
        subtype: false,
        time_created: "1522036284",
        // getAvatarSource: () => {
        //   return {
        //     source:'http://thisisaurl'
        //   }
        // }
      },
      shouldBeBlured: jest.fn(),
      message:"Message",
      title:'TITLE',
      owner_guid:"824853017709780997",
      parent_guid:"838106762591510528",
      perma_url:false,
      thumbnail_src:false,
      type:"activity",
      wire_totals: {
        tokens: 1000000000000000000
      },
      containerObj: container
      // _list: {
      //   viewed: {
      //     viewed: new Map([["1019155171608096768",true]]),
      //     addViewed: () => {
      //       return;
      //     }
      //   }
      // },
      // getThumbSource: () => {
      //   return {
      //     source:'http://thisisaurl'
      //   }
      // }
    }
  });

  return {
    'activities': activities,
    'load-next': 'aaaaaa',
    'load-previous': 'aaaaaa'
  }

}

/**
 * comment service data faker
 * @param {integer} count
 */
export function activitiesServiceFaker() {
  return { load: load}
}

