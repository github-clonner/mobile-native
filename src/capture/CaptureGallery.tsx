//@ts-nocheck
import React, { PureComponent } from 'react';

import {
  StyleSheet,
  TouchableOpacity,
  Image,
  Platform,
  FlatList,
} from 'react-native';

import CameraRoll from '@react-native-community/cameraroll';

import Icon from 'react-native-vector-icons/Ionicons';

import CenteredLoading from '../common/components/CenteredLoading';
import androidPermissionsService from '../common/services/android-permissions.service';
import logService from '../common/services/log.service';
import { CommonStyle } from '../styles/Common';
import { View } from 'react-native-animatable';
import ThemedStyles from '../styles/ThemedStyles';
/**
 * Gallery View
 */
export default class CaptureGallery extends PureComponent {
  listRef = null;

  state = {
    photos: [],
    imagesLoaded: false,
    offset: '',
    hasMore: true,
    loading: false,
  };

  static navigationOptions = {
    header: null,
    tabBarIcon: ({ tintColor }) => (
      <Icon name="md-radio-button-on" size={24} color={tintColor} />
    ),
  };

  /**
   * Load Photos
   */
  loadPhotos = async () => {
    let allowed = true;
    if (Platform.OS !== 'ios') {
      allowed = await androidPermissionsService.checkReadExternalStorage();
      if (!allowed) {
        allowed = await androidPermissionsService.readExternalStorage();
      }
    }

    if (allowed === true) {
      this._loadPhotos();
    }
  };

  /**
   * Load photos
   */
  _loadPhotos = async () => {
    if (this.state.loading || !this.state.hasMore) {
      return;
    }

    const params = {
      first: 30,
      assetType: 'All',
    };

    this.setState({ loading: true });

    if (Platform.OS === 'ios') {
      params.groupTypes = 'All';
    }
    if (this.state.offset) {
      params.after = this.state.offset;
    }

    try {
      const result = await CameraRoll.getPhotos(params);

      this.setState({
        imagesLoaded: true,
        photos: this.state.photos.concat(result.edges),
        offset: result.page_info.end_cursor,
        hasMore: result.page_info.has_next_page,
        loading: false,
      });
    } catch (err) {
      logService.exception('[CaptureGallery] loadPhotos', err);
      this.setState({ loading: false });
    }
  };

  /**
   * on component mount load photos
   */
  componentDidMount() {
    setTimeout(() => {
      this.loadPhotos();
    }, 50);
  }

  /**
   * Render
   */
  render() {
    return (
      <View
        style={[
          ThemedStyles.style.backgroundSecondary,
          CommonStyle.flexContainer,
        ]}>
        {this.state.imagesLoaded ? (
          <FlatList
            ref={this.setListRef}
            ListHeaderComponent={this.props.header}
            data={this.state.photos}
            renderItem={this.renderTile}
            style={[
              CommonStyle.flexContainer,
              ThemedStyles.style.backgroundSecondary,
            ]}
            onEndReached={this._loadPhotos}
            ListFooterComponent={
              this.state.loading ? <CenteredLoading /> : null
            }
            numColumns={3}
          />
        ) : (
          <CenteredLoading />
        )}
      </View>
    );
  }

  /**
   * Sets List reference
   */
  setListRef = (ref) => (this.listRef = ref);

  /**
   * render list tile
   */
  renderTile = (item) => {
    const node = item.item.node;

    const icon = node.type.startsWith('video') ? (
      <View style={[CommonStyle.positionAbsolute, CommonStyle.centered]}>
        <Icon name="ios-play-circle" size={24} style={CommonStyle.colorWhite} />
      </View>
    ) : null;

    return (
      <TouchableOpacity
        style={styles.tileImage}
        key={item.index}
        onPress={() => {
          this.onSelected({
            uri: node.image.uri,
            type: node.type,
            fileName: node.image.filename,
            duration: node.image.playableDuration,
            width: node.image.width,
            height: node.image.height,
          });
        }}
        testID={`GalleryImage${item.index}`}>
        <Image
          source={{ uri: node.image.uri }}
          style={styles.tileImage}></Image>
        {icon}
      </TouchableOpacity>
    );
  };

  /**
   * On media selected
   */
  onSelected = (response) => {
    // scroll to top on selection
    this.listRef.scrollToOffset({ x: 0, y: 0, animated: true });

    this.props.onSelected(response);
  };
}

const styles = StyleSheet.create({
  tileImage: {
    minHeight: 120,
    flex: 1,
    padding: 1,
  },
});
