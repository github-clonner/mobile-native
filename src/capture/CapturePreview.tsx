//@ts-nocheck
import React, { PureComponent } from 'react';
import { StyleSheet, View, Image, Platform } from 'react-native';

import MindsVideo from '../media/MindsVideo';

/**
 * Capture preview
 */
export default class CapturePreview extends PureComponent {
  /**
   * Render
   */
  render() {
    let body = null;
    switch (this.props.type) {
      case 'image/gif':
      case 'image/jpeg':
      case 'image':
      default:
        body = (
          <Image
            resizeMode="contain"
            source={{ uri: this.props.uri }}
            style={styles.preview}
          />
        );
        break;
      case 'video/mp4':
      case 'video/quicktime':
      case 'video/x-m4v':
      case 'video':
        body = (
          <View style={styles.preview}>
            <MindsVideo video={{ uri: this.props.uri }} />
          </View>
        );
        break;
    }

    return <View style={styles.wrapper}>{body}</View>;
  }
}

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'stretch',
    backgroundColor: 'black',
  },
  preview: {
    flex: 1,
  },
});
