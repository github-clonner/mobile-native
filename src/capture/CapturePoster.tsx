//@ts-nocheck
import React, { Component } from 'react';
import { StyleSheet, View, ScrollView, Text, Alert } from 'react-native';

import { observer, inject } from 'mobx-react';
import { Icon } from 'react-native-elements';
import { CommonActions } from '@react-navigation/native';

import HashtagService from '../common/services/hashtag.service';

import CaptureGallery from './CaptureGallery';
import CapturePreview from './CapturePreview';

import CaptureMetaPreview from './CaptureMetaPreview';
import CapturePostButton from './CapturePostButton';
import { CommonStyle as CS } from '../styles/Common';
import CapturePosterFlags from './CapturePosterFlags';
import UserAutocomplete from '../common/components/UserAutocomplete';
import Activity from '../newsfeed/activity/Activity';
import BlogCard from '../blogs/BlogCard';
import ActivityModel from '../newsfeed/ActivityModel';
import logService from '../common/services/log.service';
import i18n from '../common/services/i18n.service';
import settingsStore from '../settings/SettingsStore';
import CaptureTabs from './CaptureTabs';

// workaround for android copy/paste
import TextInput from '../common/components/TextInput';
import ThemedStyles from '../styles/ThemedStyles';
import remoteAction from '../common/RemoteAction';

@inject('user', 'capture', 'newsfeed')
@observer
class CapturePoster extends Component {
  /**
   * State
   */
  state = {
    postImageUri: '',
    mature: false, // @deprecated
    nsfw: [],
    share: {},
    lock: null,
    selection: {
      start: 0,
      end: 0,
    },
    time_created: null,
  };

  /**
   * On component did mount
   */
  componentDidMount() {
    let { params } = this.props.route;

    this.props.capture.reset();

    if (params) {
      if (params.text) {
        this.setText(params.text);
      } else if (params.image) {
        this.onAttachedMedia({
          type: 'image/jpeg',
          uri: params.image,
        });
      } else if (params.video) {
        this.onAttachedMedia({
          type: 'video/mp4',
          uri: params.video,
        });
      }
    }

    const { setOptions } = this.props.navigation;
    if (!params) {
      params = {};
    }
    setOptions({
      headerHideBackButton: false,
      headerRight: () => (
        <CapturePostButton
          onPress={() => (!params.isRemind ? this.submit() : this.remind())}
          text={
            params.isRemind
              ? i18n.t('capture.remind').toUpperCase()
              : i18n.t('capture.post').toUpperCase()
          }
          testID="CapturePostButton"
        />
      ),
    });

    this.loadNsfwFromPersistentStorage();
  }

  /**
   * Load last saved nsfw values
   */
  async loadNsfwFromPersistentStorage() {
    this.setState({
      nsfw: settingsStore.creatorNsfw,
    });
  }

  /**
   * On component will unmount
   */
  componentWillUnmount() {
    this.props.capture.embed.clearRichEmbedAction();
    this.deleteAttachment();
  }

  /**
   * Show context
   */
  showContext() {
    let group = this.props.route.params ? this.props.route.params.group : null;
    return group ? (
      <Text style={styles.title}>
        {' '}
        {i18n.t('capture.postingIn', { group: group.name })}{' '}
      </Text>
    ) : null;
  }

  /**
   * Nav to group
   */
  navToPrevious(entity, group) {
    const { goBack, dispatch } = this.props.navigation;
    const { params } = this.props.route;

    const activity = ActivityModel.checkOrCreate(entity);

    this.props.newsfeed.prepend(activity);

    if (params && params.parentKey) {
      const routeParams = {
        prepend: activity,
      };

      if (group) {
        routeParams.group = group;
      }

      // this.props.navigation.navigate(params.previous, routeParams);
      dispatch({
        ...CommonActions.setParams(routeParams),
        source: params.parentKey, // passed from index
      });
    }

    goBack(null);
  }

  /**
   * On tag selected in the autocomplete
   */
  onSelectTag = (text) => {
    this.setText(text);
  };

  /**
   * Set the state with cursor position
   */
  onSelectionChanges = (event) => {
    this.setState({ selection: event.nativeEvent.selection });
  };

  /**
   * Get header
   *
   * @param {boolean} showAttachmentFeatures
   */
  getHeader(showAttachmentFeatures = false) {
    return (
      <React.Fragment>
        {this.showContext()}
        <View style={styles.posterWrapper}>
          <TextInput
            style={[styles.poster, ThemedStyles.style.colorPrimaryText]}
            editable={true}
            placeholder={i18n.t('capture.placeholder')}
            placeholderTextColor={ThemedStyles.getColor('secondary_text')}
            underlineColorAndroid="transparent"
            onChangeText={this.setText}
            textAlignVertical="top"
            value={this.props.capture.text}
            multiline={true}
            selectTextOnFocus={false}
            onSelectionChange={this.onSelectionChanges}
            testID="PostInput"
          />
        </View>
        {showAttachmentFeatures && this.getAttachFeature()}
      </React.Fragment>
    );
  }

  /**
   * Render
   */
  render() {
    const params = this.props.route.params || {};

    return params.isRemind ? this.renderRemind() : this.renderNormal();
  }

  /**
   * Screen content for poster
   */
  renderNormal() {
    const navigation = this.props.navigation;

    const params = this.props.route.params || {};

    return (
      <View style={[CS.flexContainer, ThemedStyles.style.backgroundSecondary]}>
        <CaptureGallery
          onSelected={this.onAttachedMedia}
          header={this.getHeader(true)}
        />
        <UserAutocomplete
          text={this.props.capture.text}
          selection={this.state.selection}
          onSelect={this.onSelectTag}
        />
      </View>
    );
  }

  /**
   * Screen content for remind
   */
  renderRemind() {
    const text = this.props.capture.text;

    return (
      <View style={[CS.flexContainer, ThemedStyles.style.backgroundSecondary]}>
        <ScrollView
          style={styles.posterAndPreviewWrapper}
          keyboardShouldPersistTaps={'always'}
          removeClippedSubviews={false}>
          {this.getHeader()}
          {this.getRemind()}
        </ScrollView>
        <UserAutocomplete
          text={text}
          selection={this.state.selection}
          onSelect={this.onSelectTag}
        />
      </View>
    );
  }

  /**
   * Get remind card
   */
  getRemind() {
    const { params } = this.props.route;
    const ShowComponent =
      params.entity.subtype === 'blog' ? BlogCard : Activity;
    return (
      <ShowComponent
        hideTabs={true}
        entity={params.entity}
        navigation={this.props.navigation}
      />
    );
  }

  /**
   * Get attachment feature
   */
  getAttachFeature() {
    const attachment = this.props.capture.attachment;
    return (
      <React.Fragment>
        {(this.props.capture.embed.meta ||
          this.props.capture.embed.metaInProgress) && (
          <CaptureMetaPreview
            meta={this.props.capture.embed.meta}
            inProgress={this.props.capture.embed.metaInProgress}
            onRemove={this.props.capture.embed.clearRichEmbedAction}
          />
        )}

        <CapturePosterFlags
          containerStyle={[CS.rowJustifyEnd]}
          matureValue={this.state.mature}
          hideShare={true}
          shareValue={this.state.share}
          lockValue={this.state.lock}
          nsfwValue={this.state.nsfw}
          timeCreatedValue={this.state.time_created}
          onMature={this.onMature}
          onNsfw={this.onNsfw}
          onShare={this.onShare}
          onLocking={this.onLocking}
          onScheduled={this.onScheduled}
        />

        {attachment.hasAttachment && (
          <View style={styles.preview}>
            <CapturePreview uri={attachment.uri} type={attachment.type} />
            <Icon
              raised
              reverse
              name="md-close"
              type="ionicon"
              color="#4690DF"
              size={18}
              containerStyle={styles.deleteAttachment}
              onPress={() => this.deleteAttachment()}
              testID="AttachmentDeleteButton"
            />
          </View>
        )}
        <CaptureTabs onSelectedMedia={this.onAttachedMedia} />
      </React.Fragment>
    );
  }

  /**
   * Attach Media
   */
  onAttachedMedia = async (response) => {
    const attachment = this.props.capture.attachment;
    let group = this.props.route.params ? this.props.route.params.group : null;
    let extra = null;

    if (group) {
      extra = { container_guid: group.guid };
    }

    try {
      const result = await attachment.attachMedia(response, extra);
    } catch (err) {
      logService.exception(err);
      Alert.alert(i18n.t('capture.uploadError'));
    }
  };

  /**
   * Delete attachment
   */
  async deleteAttachment() {
    const attachment = this.props.capture.attachment;
    // delete only if it has an attachment
    if (attachment.hasAttachment) {
      const result = await attachment.delete();
      if (result === false) Alert.alert(i18n.t('capture.errorDeleting'));
    }
  }

  /**
   * Create a remind
   */
  async remind() {
    const { params } = this.props.route;
    const message = this.props.capture.text;
    const metadata = params.entity.getClientMetadata();

    const post = {
      message,
      ...metadata,
    };

    let group = this.props.route.params ? this.props.route.params.group : null;

    // if hashtag count greater than 5
    if (HashtagService.slice(message).length > HashtagService.maxHashtags) {
      Alert.alert(
        i18n.t('capture.maxHashtags', {
          maxHashtags: HashtagService.maxHashtags,
        }),
      );
      return false;
    }

    return await remoteAction(async () => {
      const response = await this.props.capture.remind(
        params.entity.guid,
        post,
      );
      this.navToPrevious(response.entity, group);
    });
  }

  /**
   * Submit post
   */
  async submit() {
    const attachment = this.props.capture.attachment;
    const text = this.props.capture.text;

    if (attachment.hasAttachment && attachment.uploading) {
      Alert.alert(i18n.t('capture.pleaseTryAgain'));
      return false;
    }

    if (
      !attachment.hasAttachment &&
      !text &&
      (!this.props.capture.embed.meta || !this.props.capture.embed.meta.url)
    ) {
      Alert.alert(i18n.t('capture.nothingToPost'));
      return false;
    }

    if (HashtagService.slice(text).length > HashtagService.maxHashtags) {
      //if hashtag count greater than 5
      Alert.alert(
        i18n.t('capture.maxHashtags', {
          maxHashtags: HashtagService.maxHashtags,
        }),
      );
      return false;
    }

    let newPost = {
      message: text,
      wire_threshold: this.state.lock,
      time_created: this.formatTimeCreated(),
    };

    newPost.nsfw = this.state.nsfw || [];

    if (attachment.guid) {
      newPost.attachment_guid = attachment.guid;
      newPost.attachment_license = attachment.license;
    }

    for (let network in this.state.share) {
      if (this.state.share[network]) {
        newPost[network] = 1;
      }
    }

    if (this.props.capture.embed.meta) {
      newPost = Object.assign(newPost, this.props.capture.embed.meta);
    }

    if (this.props.route.params && this.props.route.params.group) {
      newPost.container_guid = this.props.route.params.group.guid;
    }

    if (this.props.capture.tags && this.props.capture.tags.length) {
      newPost.tags = this.props.capture.allTags;
    }

    return await remoteAction(async () => {
      let response = await this.props.capture.post(newPost);

      if (this.props.reset) {
        this.props.reset();
      }

      // clear attachment data
      attachment.clear();

      this.setState({
        meta: null,
        mature: false,
        nsfw: [],
        share: {},
        lock: null,
      });
      this.props.capture.setText('');

      if (this.props.onComplete) {
        this.props.onComplete(response.entity);
      } else if (this.props.route.params && this.props.route.params.group) {
        this.navToPrevious(response.entity, this.props.route.params.group);
      } else {
        this.navToPrevious(response.entity);
      }
      return response;
    });
  }

  /**
   * Set text
   * @param {string} text
   */
  setText = (text) => {
    this.props.capture.setText(text);
    this.props.capture.embed.richEmbedCheck(text);
  };

  /**
   * On mature value change
   */
  onMature = () => {
    const mature = !this.state.mature;
    this.setState({ mature });
  };

  /**
   * On nsfw value change
   */
  onNsfw = (values) => {
    const nsfw = [...values];
    this.setState({ nsfw });
  };

  onShare = (network) => {
    const share = Object.assign({}, this.state.share);

    if (share[network]) {
      delete share[network];
    } else {
      share[network] = true;
    }

    this.setState({ share });
  };

  onLocking = (lock) => {
    this.setState({ lock });
  };

  onScheduled = (timeCreated) => {
    this.setState({ time_created: timeCreated });
  };

  formatTimeCreated = () => {
    let time_created;
    if (this.state.time_created) {
      time_created = new Date(this.state.time_created).getTime();
    } else {
      time_created = Date.now();
    }
    return Math.floor(time_created / 1000);
  };
}

export default CapturePoster;

const styles = StyleSheet.create({
  posterAndPreviewWrapper: {
    flex: 1,
  },
  posterWrapper: {
    minHeight: 100,
    flexDirection: 'row',
  },
  title: {
    margin: 0,
    paddingHorizontal: 10,
  },
  poster: {
    alignContent: 'flex-start',
    padding: 15,
    paddingTop: 15,
    flex: 1,
  },
  preview: {
    flex: 1,
    minHeight: 200,
    flexDirection: 'row',
    alignItems: 'stretch',
    position: 'relative',
  },
  gallery: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'stretch',
  },
  deleteAttachment: {
    position: 'absolute',
    right: 8,
    top: 0,
  },
});
