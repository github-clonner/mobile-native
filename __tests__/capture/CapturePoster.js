import 'react-native';
import React from 'react';
import { Alert } from 'react-native';
import { shallow } from 'enzyme';
import { Icon } from 'react-native-elements';
import CameraRoll from '@react-native-community/cameraroll';

import CapturePoster from '../../src/capture/CapturePoster';
import CapturePreview from '../../src/capture/CapturePreview';
import CaptureGallery from '../../src/capture/CaptureGallery';
import UserStore from '../../src/auth/UserStore';
import CaptureStore from '../../src/capture/CaptureStore';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { getPhotosFaker } from '../../__mocks__/fake/CameraRollFaker';

jest.mock('../../src/auth/UserStore');
jest.mock('../../src/capture/CaptureStore');
jest.mock('../../src/capture/CapturePostButton', () => 'CapturePostButton');
jest.mock('../../src/capture/CapturePreview', () => 'CapturePreview');
jest.mock('../../src/capture/CapturePosterFlags', () => 'CapturePosterFlags');
jest.mock('../../src/common/services/translation.service');
jest.mock('@react-navigation/native');

Alert.alert = jest.fn();
CameraRoll.getPhotos = jest.fn();
// fake camera roll data
const response = getPhotosFaker(5);
CameraRoll.getPhotos.mockResolvedValue(response);

/**
 * Tests
 */
describe('cature poster component', () => {
  let userStore, capture;

  const newsfeed = {
    prepend: jest.fn()
  };

  const navigation = {
    navigate: jest.fn(),
    dispatch: jest.fn(),
    setOptions: jest.fn(),
    goBack: jest.fn(),
  };
  const route = { params: {} };

  const paramsVideo = { uri: 'file://video.mp4', type: 'video/mp4' };

  const paramsImage = { uri: 'file://image.jpg', type: 'image/jpeg' };

  beforeEach(() => {
    userStore = new UserStore();
    capture = new CaptureStore();
    capture.attachment.attachMedia.mockClear();
    navigation.navigate.mockClear();
    navigation.dispatch.mockClear();
    navigation.setOptions.mockClear();
    capture.post.mockClear();
  });

  it('should renders correctly', () => {
    const screen = renderer
      .create(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      )
      .toJSON();

    expect(screen).toMatchSnapshot();
  });

  it('should receive text parameters on did mount', () => {
    route.params = { text: 'hello' };

    const wrapper = shallow(
      <CapturePoster.wrappedComponent
        user={userStore}
        capture={capture}
        navigation={navigation}
        route={route}
      />,
    );

    // should be called
    expect(capture.setText).toHaveBeenCalled();

    // should be called with text hello
    expect(capture.setText).toHaveBeenCalledWith('hello');
  });

  it('should receive image parameters on did mount and attach it', () => {
    route.params = { image: 'file://image.jpg' };

    const wrapper = shallow(
      <CapturePoster.wrappedComponent
        user={userStore}
        capture={capture}
        navigation={navigation}
        route={route}
      />,
    );

    // should be called
    expect(capture.attachment.attachMedia).toHaveBeenCalled();

    // should be called only once
    expect(capture.attachment.attachMedia.mock.calls.length).toBe(1);

    // should be called with the image
    expect(capture.attachment.attachMedia.mock.calls[0][0]).toEqual({
      type: 'image/jpeg',
      uri: 'file://image.jpg',
    });
  });

  it('should receive video parameters on did mount and attach it', () => {
    route.params = { video: 'file://video.mp4' };

    const wrapper = shallow(
      <CapturePoster.wrappedComponent
        user={userStore}
        capture={capture}
        navigation={navigation}
        route={route}
      />,
    );

    // should be called
    expect(capture.attachment.attachMedia).toHaveBeenCalled();

    // should be called only once
    expect(capture.attachment.attachMedia.mock.calls.length).toBe(1);

    // should be called with the video
    expect(capture.attachment.attachMedia.mock.calls[0][0]).toEqual({
      type: 'video/mp4',
      uri: 'file://video.mp4',
    });
  });

  it('should show the preview when an image is attached', async done => {
    try {
      // emulate image attachment
      capture.attachment.hasAttachment = true;
      capture.attachment.uri = paramsImage.uri;
      capture.attachment.type = paramsImage.type;

      const wrapper = renderer.create(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      const gallery = wrapper.root.findByType(CaptureGallery);

      await gallery.instance._loadPhotos();

      // find Capture Preview
      const preview = wrapper.root.findByType(CapturePreview);

      expect(preview).toBeDefined();

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it('should show the preview when a video is attached', async done => {
    try {
      // emulate video attachment
      capture.attachment.hasAttachment = true;
      capture.attachment.uri = paramsVideo.uri;
      capture.attachment.type = paramsVideo.type;

      const wrapper = renderer.create(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      const gallery = wrapper.root.findByType(CaptureGallery);

      await gallery.instance._loadPhotos();

      // find Capture Preview
      const preview = wrapper.root.findByType(CapturePreview);

      expect(preview).toBeDefined();

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it('should call attachment store delete when user tap delete', async done => {
    try {
      // emulate video attachment
      capture.attachment.hasAttachment = true;
      capture.attachment.uri = paramsVideo.uri;
      capture.attachment.type = paramsVideo.type;

      const wrapper = renderer.create(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      const gallery = wrapper.root.findByType(CaptureGallery);

      await gallery.instance._loadPhotos();

      // find delete icon
      const icon = wrapper.root.findByType(Icon);

      expect(icon).toBeDefined();

      // simulate press on image
      icon.props.onPress();

      // should be called
      expect(capture.attachment.delete).toHaveBeenCalled();

      // should be called only once
      expect(capture.attachment.delete.mock.calls.length).toBe(1);

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it('should call attachment store delete when user leave', async done => {
    try {
      // emulate video attachment
      capture.attachment.hasAttachment = true;
      capture.attachment.uri = paramsVideo.uri;
      capture.attachment.type = paramsVideo.type;

      const wrapper = shallow(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      // unmount
      await wrapper.instance().componentWillUnmount();

      // should be called
      expect(capture.attachment.delete).toHaveBeenCalled();

      // should be called only once
      expect(capture.attachment.delete.mock.calls.length).toBe(1);

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it('should return false if the user post before finish an upload', async done => {
    try {
      // emulate video attachment
      capture.attachment.hasAttachment = true;
      capture.attachment.uri = paramsVideo.uri;
      capture.attachment.type = paramsVideo.type;
      capture.attachment.uploading = true;

      const wrapper = shallow(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      // submit
      const result = await wrapper.instance().submit();

      // should be false
      expect(result).toBe(false);

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it('should return false if there is nothing to post', async done => {
    try {
      const wrapper = shallow(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      // submit
      const result = await wrapper.instance().submit();

      // should be false
      expect(result).toBe(false);

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it("should call store's post on submit and return the response", async done => {
    try {
      const wrapper = shallow(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      wrapper.instance().setText('some awesome post');

      const response = { entity: { guid: 1, title: 'some awesome post' } };

      capture.post.mockResolvedValue(response);

      // submit
      const result = await wrapper.instance().submit();

      // should be called
      expect(capture.post).toHaveBeenCalled();

      // should be called only once
      expect(capture.post.mock.calls.length).toBe(1);

      const entity = capture.post.mock.calls[0][0];
      expect(capture.post.mock.calls[0][0]).toEqual({
        nsfw: [],
        message: 'some awesome post',
        wire_threshold: null,
        time_created: entity.time_created,
      });

      expect(result).toEqual(response);

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it('should call onComplete after submit', async done => {
    try {
      const onCompleteMock = jest.fn();

      const wrapper = shallow(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          onComplete={onCompleteMock}
        />,
      );

      wrapper.instance().setText('some awesome post');

      const response = { entity: { guid: 1, title: 'some awesome post' } };

      capture.post.mockResolvedValue(response);

      // submit
      const result = await wrapper.instance().submit();

      // should be called
      expect(onCompleteMock).toHaveBeenCalled();

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it('should add third party networks share to post', async done => {
    try {
      const wrapper = shallow(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      wrapper.instance().setText('some awesome post');
      wrapper.instance().onShare('facebook');
      wrapper.instance().onShare('twitter');

      const response = { entity: { guid: 1, title: 'some awesome post' } };

      capture.post.mockResolvedValue(response);

      // submit
      const result = await wrapper.instance().submit();

      // should be called
      expect(capture.post).toHaveBeenCalled();

      // should be called only once
      expect(capture.post.mock.calls.length).toBe(1);

      const entity = capture.post.mock.calls[0][0];
      expect(capture.post.mock.calls[0][0]).toEqual({
        nsfw: [],
        message: 'some awesome post',
        wire_threshold: null,
        facebook: 1,
        twitter: 1,
        time_created: entity.time_created,
      });

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it("should call store's post with attachment guid", async done => {
    try {
      const wrapper = shallow(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      // emulate image attachment
      capture.attachment.hasAttachment = true;
      capture.attachment.guid = 1000;
      capture.attachment.uri = paramsImage.uri;
      capture.attachment.type = paramsImage.type;
      capture.attachment.uploading = false;

      wrapper.instance().setText('some awesome post');

      const response = { entity: { guid: 1, title: 'some awesome post' } };

      capture.post.mockResolvedValue(response);

      // submit
      const result = await wrapper.instance().submit();

      console.log('results',result)

      // should be called
      expect(capture.post).toHaveBeenCalled();

      // should be called only once
      expect(capture.post).toBeCalled();

      const entity = capture.post.mock.calls[0][0];

      // should send the attachment data
      expect(capture.post.mock.calls[0][0]).toEqual({
        nsfw: [],
        message: 'some awesome post',
        wire_threshold: null,
        attachment_guid: 1000,
        attachment_license: '',
        time_created: entity.time_created,
      });

      // should return server response
      expect(result).toEqual(response);

      // should clear the attachment
      expect(capture.attachment.clear).toHaveBeenCalled();

      done();
    } catch (e) {
      done.fail(e);
    }
  });

  it('should alert the user on submit error', async done => {
    try {
      const wrapper = shallow(
        <CapturePoster.wrappedComponent
          user={userStore}
          capture={capture}
          navigation={navigation}
          route={route}
          newsfeed={newsfeed}
        />,
      );

      wrapper.instance().setText('some awesome post');

      capture.post.mockResolvedValue(new Error('some error'));

      // submit
      const result = await wrapper.instance().submit();

      // should be called
      expect(Alert.alert).toHaveBeenCalled();

      done();
    } catch (e) {
      done.fail(e);
    }
  });
});
