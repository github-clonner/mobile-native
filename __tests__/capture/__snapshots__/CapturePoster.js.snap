// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cature poster component should renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flex": 1,
          "justifyContent": "center",
          "minHeight": 100,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </View>
</View>
`;
